File: src/main/java/com/barbulescu/codedojo/exercise0003/Money.java
package com.barbulescu.codedojo.exercise0003;

import java.math.BigDecimal;

public record Money(BigDecimal amount, String currency) {
    public Money {
        if (amount == null) {
            throw new IllegalArgumentException("amount must not be null");
        }
        if (amount.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("amount must not be negative");
        }
        if (currency == null || currency.isBlank()) {
            throw new IllegalArgumentException("currency must not be blank");
        }
    }

    public Money plus(Money other) {
        ensureSameCurrency(other);
        return new Money(amount.add(other.amount), currency);
    }

    public Money multiply(int multiplier) {
        if (multiplier < 0) {
            throw new IllegalArgumentException("multiplier must not be negative");
        }
        return new Money(amount.multiply(BigDecimal.valueOf(multiplier)), currency);
    }

    private void ensureSameCurrency(Money other) {
        if (other == null) {
            throw new IllegalArgumentException("other must not be null");
        }
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException("Cannot combine " + currency + " with " + other.currency);
        }
    }
}

File: src/main/kotlin/com/barbulescu/codedojo/exercise0003/kotlin/Money.kt
package com.barbulescu.codedojo.exercise0003.kotlin

import java.math.BigDecimal

data class Money(val amount: BigDecimal, val currency: NonBlankString) {

    init {
        require(amount >= BigDecimal.ZERO) { "Amount must not be negative" }
    }

    operator fun plus(other: Money): Money {
        ensureSameCurrency(other)
        return Money(amount + other.amount, currency)
    }

    operator fun minus(other: Money): Money {
        ensureSameCurrency(other)
        return Money(amount - other.amount, currency)
    }

    private fun ensureSameCurrency(other: Money) {
        require(currency == other.currency) { """Cannot combine $currency with ${other.currency}""" }
    }

    fun multiply(multiplier: Int): Money {
        require(multiplier >= 0) { "multiplier must not be negative" }
        return Money(amount.multiply(multiplier.toBigDecimal()), currency)
    }
}
